### PR TITLE
Sandbox

### PR DESCRIPTION
**Pull request by smarches**
Summary of changes:

- initial implementation of Circle class with tests
- update README to provide more explanation and usage examples
- improve efficiency and simplicity of fast_encode and coverage functions
- added some helper/utility methods that can be shared across shapes
- addressed encoding issue from pygeohash with randomized test to ensure correctness, see https://github.com/vinsci/geohash/pull/6 and https://github.com/vinsci/geohash/issues/9

### CODE REVIEW

File: neathgeohash/geohash.py [hunk 1]
```diff
@@ -116,58 +110,68 @@ def encode(latitude, longitude, precision=12):
...
+#     c3 = np.uint8(np.bitwise_and(np.right_shift([g], 40), __mask))[0]
+#     c2 = np.uint8(np.bitwise_and(np.right_shift([g], 45), __mask))[0]
+#     c1 = np.uint8(np.bitwise_and(np.right_shift([g], 50), __mask))[0]
+#     c0 = np.uint8(np.bitwise_and(np.right_shift([g], 55), __mask))[0]
+#     return np.column_stack((c0, c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11))
+
+# b64_nums = [12345,84795434,84795400]
+# encode_quintets(b64_nums)
+# __encode_base32(b64_nums)
+
```
Comment (mpdwulit) [inline neathgeohash/geohash.py]: Please remove commented out code. We really don't need it here (it is in the git).

File: neathgeohash/__init__.py [hunk 2]
```diff
@@ -8,16 +8,26 @@
...
-from .Ellipse import Ellipse
+from .geohash import decode, decode_exactly, decode_into_corners, encode, fast_encode
+from .geohash import get_neighbors, generate_children, get_neighbor_collection
+from .geohash import latlon_offset, atomize_geohash
 from .Line import Line
-from .geohash import decode_exactly, decode, encode, fast_encode
+from .Ellipse import Ellipse
+from .Circle import Circle
 
 __author__ = 'mpdwulit'
```
Comment (mpdwulit) [inline neathgeohash/__init__.py:18]: Please remove the author from the code,
at this point we both are authors ;)

File: neathgeohash/geohash.py [hunk 3]
```diff
@@ -46,50 +39,133 @@ def decode_exactly(geohash):
...
+#             else:      # adds latitude info
+#                 lat_err /= 2
+#                 if cd & mask:
+#                     lat_interval = ((lat_interval[0]+lat_interval[1])/2, lat_interval[1])
+#                 else:
+#                     lat_interval = (lat_interval[0], (lat_interval[0]+lat_interval[1])/2)
+#             is_even = not is_even
+#     lat = (lat_interval[0] + lat_interval[1]) / 2
+#     lon = (lon_interval[0] + lon_interval[1]) / 2
+#     return lat, lon, lat_err, lon_err
```
Comment (mpdwulit) [inline neathgeohash/geohash.py:114]: Please remove junk code ;). At this point, we don't need the old code to pollute.

File: neathgeohash/geohash.py [hunk 4]
```diff
@@ -46,50 +39,133 @@ def decode_exactly(geohash):
...
+    Truncates/pads a float f to n decimal places without rounding
+    see https://stackoverflow.com/questions/783897/truncating-floats-in-python
+    '''
+    if n <= 0: return str(int(f))
+    s = '{}'.format(f)
+    if 'e' in s or 'E' in s:
+        return '{0:.{1}f}'.format(f, n)
+    i, _, d = s.partition('.')
+    return f"{i}.{(d+'0'*n)[:n]}"
+    # return '.'.join([i, (d+'0'*n)[:n]])
```
Comment (mpdwulit) [inline neathgeohash/geohash.py:127]: Please don't keep old code

File: neathgeohash/geohash.py [hunk 5]
```diff
@@ -46,50 +39,133 @@ def decode_exactly(geohash):
...
-    # Format to the number of decimals that are known
-    lats = "%.*f" % (max(1, int(round(-log10(lat_err)))) - 1, lat)
-    lons = "%.*f" % (max(1, int(round(-log10(lon_err)))) - 1, lon)
-    if '.' in lats: lats = lats.rstrip('0')
-    if '.' in lons: lons = lons.rstrip('0')
-    return lats, lons
+    lat_n = ceil(-log10(lat_err))
+    lon_n = ceil(-log10(lon_err))
+    res = truncate(lat,lat_n), truncate(lon,lon_n)
+    # see https://github.com/vinsci/geohash/issues/9,
```
Comment (mpdwulit) [inline neathgeohash/geohash.py:138]: Do we have a unit test for this case?

File: tests/performance_cmp.py [hunk 6]
```diff
@@ -27,14 +29,16 @@ def test_run_encode(self):
...
-        ngh.fast_encode(loc_list)
+        ngh.fast_encode(np.asarray(loc_list))
         pr.disable()
         s = io.StringIO()
         ps = pstats.Stats(pr, stream=s).strip_dirs().sort_stats('cumtime')
         ps.print_stats()
         print(s.getvalue())
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (mpdwulit) [inline tests/performance_cmp.py:44]: Add the new line.

File: tests/test_Circle.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            np.random.uniform(-160,160),
+            np.random.uniform(1000,10000)
+        )
+        with self.assertRaises(ValueError):
+            c0.geohash_border(depth = 100)
+        # check of depth estimation process:
+        b1 = c0.geohash_border(depth = 10,max_gh = 1000)
+        self.assertTrue(all(len(b) < 10 for b in b1))
+
+    def test_border_and_coverage(self):
```
Comment (mpdwulit) [inline tests/test_Circle.py:52]: The unit test has to test only one thing. Split this in to two independent tests.

File: tests/test_Circle.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        self.assertTrue(p1 in c0)
+        self.assertTrue((10.001,19.998) in c0)
+        self.assertFalse((8,17) in c0)
+
+    def test_point_count(self):
+        c0 = ngh.Circle(10,20,3)
+        self.assertEqual(c0.overlaps_geohash('s3y0zh7w1z'),4)
+        self.assertEqual(c0.overlaps_geohash('s3y0zh7w1'), 1)
+    
+    # purely for code coverage
```
Comment (mpdwulit) [inline tests/test_Circle.py:26]: Ok. This comment is a red flag for me. Either we testing code which does something useful or we have useless code? Which one is it? Either remove the comment or remove code

What you are doing here is checking if the representation is right. I recommend removing the comment.

File: tests/test_geohash.py [hunk 9]
```diff
@@ -16,34 +18,64 @@ def test_encode(self):
         self.assertEqual('ezs42e44yx96', gh)
 
     def test_decode(self):
-        ll = ngh.decode('ezs42e44yx96')
-        self.assertEqual(('42.6', '-5.6'), ll)
+        b32 = list('0123456789bcdefghjkmnpqrstuvwxyz')
```
Comment (mpdwulit) [inline tests/test_geohash.py:21]: Thank you for that one. Great job!

File: tests/test_geohash.py [hunk 10]
```diff
@@ -16,34 +18,64 @@ def test_encode(self):
...
         gh_fe = ngh.fast_encode([[42.6, -5.6], [42.6, -5.6], [42.6, -5.6]])
-        self.assertEqual((3,), gh_fe.shape, )
-
-    def test_fast_encode(self):
-        gh = ngh.encode(42.6, -5.6)
-        gh_fe = ngh.fast_encode([[42.6, -5.6]])
-        self.assertEqual(gh.encode(encoding='UTF-8'), gh_fe[0])
+        self.assertEqual((3,), gh_fe.shape)
 
     def test_fast_encode(self):
```
Comment (mpdwulit) [inline tests/test_geohash.py:45]: Please do not combine a few unit tests into one big test. That defeats the purpose of unit testing. 
The primary goal of a unit test is to validate one logical thing. I recommend reverting back.

The goal of the unit test is to see if there is a problem and then to provide the context for the problem.
As an author of the code, I add an edge case test and indicate what the test was testing because I want to be polite to all potential developers who will work with the code. If the unit test fails one should now where it failed (debugger) and what was the setup.

File: tests/performance_cmp.py (hunk 6 above)
Comment (mpdwulit) [inline tests/performance_cmp.py:44]: Add the new line
